tools/distpartitioning/utils.py

```diff
@@ -0,0 +1,253 @@
+    ntypes = [(key, global_nid_ranges[key][0,0]) for key in global_nid_ranges]
+    ntypes.sort(key=lambda e: e[1])
+
+    #Create node_typename -> node_type dictionary
+    ntypes = [e[0] for e in ntypes]
+    ntypes_map = {e: i for i, e in enumerate(ntypes)}
+
+    return ntypes_map, ntypes
+
+def write_metadata_json(metadata_list, output_dir, graph_name):
+    """Merge json schema's from each of the rank's on rank-0. 
+    This utility function, to be used on rank-0, to create aggregated json file.
+
+    Parameters:
+    -----------
+    metadata_list : list of json (dictionaries)
+        a list of json dictionaries to merge on rank-0
+    output_dir : string
+        output directory path in which results are stored (as a json file)
+    graph-name : string
+        a string specifying the graph name
+    """
+    #Initialize global metadata
+    graph_metadata = {}
+
+    #Merge global_edge_ids from each json object in the input list
+    edge_map = {}
+    x = metadata_list[0]["edge_map"]
+    for k in x:
+        edge_map[k] = []
+        for idx in range(len(metadata_list)):
+            edge_map[k].append(metadata_list[idx]["edge_map"][k][0])
+    graph_metadata["edge_map"] = edge_map
+
+    graph_metadata["etypes"] = metadata_list[0]["etypes"]
+    graph_metadata["graph_name"] = metadata_list[0]["graph_name"]
+    graph_metadata["halo_hops"] = metadata_list[0]["halo_hops"]
+
+    #Merge global_nodeids from each of json object in the input list
+    node_map = {}
+    x = metadata_list[0]["node_map"]
+    for k in x:
+        node_map[k] = []
+        for idx in range(len(metadata_list)):
+            node_map[k].append(metadata_list[idx]["node_map"][k][0])
+    graph_metadata["node_map"] = node_map
+
+    graph_metadata["ntypes"] = metadata_list[0]["ntypes"]
+    graph_metadata["num_edges"] = sum([metadata_list[i]["num_edges"] for i in range(len(metadata_list))])
+    graph_metadata["num_nodes"] = sum([metadata_list[i]["num_nodes"] for i in range(len(metadata_list))])
+    graph_metadata["num_parts"] = metadata_list[0]["num_parts"]
+    graph_metadata["part_method"] = metadata_list[0]["part_method"]
+
+    for i in range(len(metadata_list)):
+        graph_metadata["part-{}".format(i)] = metadata_list[i]["part-{}".format(i)]
+
+    with open('{}/{}.json'.format(output_dir, graph_name), 'w') as outfile: 
+        json.dump(graph_metadata, outfile, sort_keys=True, indent=4)
+
+def augment_edge_data(edge_data, partitions):
+    """Add partition-id (rank which owns an edge) column to the edge_data.
+    
+    Parameters:
+    -----------
+    edge_data : numpy ndarray
+        Edge information as read from the xxx_edges.txt file
+    partitions : dictionary
+        METIS assigned partitions to each of the nodes in the input graph
+
+    Returns:
+    --------
+    numpy ndarray
+        original edge_data with one additional column
+    """
+    tgtrank = [partitions[x[1]] for x in edge_data]
+    return np.c_[edge_data, tgtrank]
```

Comment:
concatenating them will require a large amount of memory copy. can this be avoided?